adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/AbstractVertxBasedCoapAdapter.java

```diff
@@ -0,0 +1,419 @@
+import org.eclipse.californium.core.server.resources.CoapExchange;
+import org.eclipse.hono.client.ClientErrorException;
+import org.eclipse.hono.client.MessageSender;
+import org.eclipse.hono.client.ServerErrorException;
+import org.eclipse.hono.service.AbstractProtocolAdapterBase;
+import org.eclipse.hono.service.auth.device.Device;
+import org.eclipse.hono.util.Constants;
+import org.eclipse.hono.util.EventConstants;
+import org.eclipse.hono.util.ResourceIdentifier;
+import org.eclipse.hono.util.TelemetryConstants;
+import org.eclipse.hono.util.TenantObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * Base class for a Vert.x based Hono protocol adapter that uses CoAP. It provides access to the Telemetry and Event
+ * API.
+ * 
+ * @param <T> The type of configuration properties used by this service.
+ */
+public abstract class AbstractVertxBasedCoapAdapter<T extends CoapAdapterProperties>
+        extends AbstractProtocolAdapterBase<T> {
+
+    protected final Logger log = LoggerFactory.getLogger(getClass());
+
+    /**
+     * COAP server. Created from blocking execution, therefore use volatile.
+     */
+    private CoapServer server;
+    private volatile Endpoint insecureEndpoint;
+    private CoapAdapterMetrics metrics;
+
+    /**
+     * Sets the metrics for this service.
+     *
+     * @param metrics The metrics
+     */
+    @Autowired
+    public final void setMetrics(final CoapAdapterMetrics metrics) {
+        this.metrics = metrics;
+    }
+
+    /**
+     * @return {@link CoAP#DEFAULT_COAP_SECURE_PORT}
+     */
+    @Override
+    public final int getPortDefaultValue() {
+        return CoAP.DEFAULT_COAP_SECURE_PORT;
+    }
+
+    /**
+     * @return {@link CoAP#DEFAULT_COAP_PORT}
+     */
+    @Override
+    public final int getInsecurePortDefaultValue() {
+        return CoAP.DEFAULT_COAP_PORT;
+    }
+
+    @Override
+    protected final int getActualPort() {
+        return Constants.PORT_UNCONFIGURED;
+    }
+
+    @Override
+    protected final int getActualInsecurePort() {
+        int port = Constants.PORT_UNCONFIGURED;
+        final Endpoint endpoint = insecureEndpoint;
+        if (endpoint != null) {
+            port = endpoint.getAddress().getPort();
+        }
+        return port;
+    }
+
+    /**
+     * Called during {@link #doStart(Future)}.
+     * 
+     * Note: The call is in the scope of "executeBlocking" and therefore special care accessing this instance must be
+     * obeyed.
+     * 
+     * @param adapterContext context of this adapter. Intended for schedule calls back into this vertical.
+     * @param server coap server to add resources.
+     */
+    protected abstract void addResources(Context adapterContext, CoapServer server);
```

Comment:
no need to pass in the adapter's vert.x context. a subclass can simply refer to the class's `context` field for that purpose because all adapters are `Verticle`s.

Reply:
>     * Note: The call is in the scope of "executeBlocking" and therefore special care accessing this instance must be
>     * obeyed.

As mentioned above, if you are sure, that accessing the `context` from other threads is save, OK.
But FMPOV, the behaviour according other threads seems to be not documented.